### DISCUSSION TITLE
Trying to make a simple MQTT broker connection (mqtt_as)

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#13232] [General]
Hello everybody.

I'm using the @peterhinch libraries and his example to make a simple connection with my "Hive MQ" broker cluster (adding Hive MQ particularities) and it doesn't work:

     from mqtt_as import MQTTClient, config
     import asyncio

    #Local configuration
    config["ssid"] = "my_ssid"  # Optional on ESP8266
    config["wifi_pw"] = "my_pass"
    config["user"] = "my_user"
    config["password"] = "my_pass"
    config["client_id"] = "a_client_id"
    broker = "xxx.s2.eu.hivemq.cloud"
    config["server"] = broker
    config["port"] = 0
    config["ssl"] = True
    config["ssl_params"] = {"server_hostname": broker , }

    def callback(topic, msg, retained):
        print((topic, msg, retained))

    async def conn_han(client):
        await client.subscribe('foo_topic', 1)

    async def main(client):
        await client.connect()
        n = 0
        while True:
            await asyncio.sleep(5)
            print('publish', n)
            # If WiFi is down the following will pause for the duration.
            await client.publish('result', '{}'.format(n), qos = 1)
            n += 1

    config['subs_cb'] = callback
    config['connect_coro'] = conn_han

    MQTTClient.DEBUG = True  # Optional: print diagnostic messages
    client = MQTTClient(config)
    try:
        asyncio.run(main(client))
    finally:
        client.close()  # Prevent LmacRxBlk:1 errors

Somebody could help me?

Thanks. Merry christmas to everyone.

**Comment by 2dof** [comment]
I usually connect to hivemq like  in example below ("My example:")  based on example/tutorial from[ot-reading-sensor-data-raspberry-pi-pico](https://www.hivemq.com/article/iot-reading-sensor-data-raspberry-pi-pico-w-micropython-mqtt-node-red/)   and then  run node-red to visualising data. 
If you subscribe then ensure  that client (broker) subscribe (send data back), I usually test first all "manually" form node-red. 

also wy  await in ``` await client.publish('result', '{}'.format(n), qos = 1)``` in loop when you have   ```await asyncio.sleep(5)``` in while loop ( ...sleep(5) should be at the end while loop)

see some other example ( not asnychio) with mosquito as broker on local machine (your laptop) (it help (at least in my case) to faster design and test code and then easy way to switch to hivemq  [micropython-mqtt-esp32-esp8266](https://www.donskytech.com/micropython-mqtt-esp32-esp8266/)


My example: 
``` python 
def connectMQTT():
    client = MQTTClient(
    client_id=b"l<client_name>",
    server=<server_adres>    #b"da375fb.........92bf6d.eu.hivemq.cloud",
    port=8883,                   # port No defined in hivemq.cloud (ussualu 8883)
    user=<user name> ,     #                   
    password=<password do server(cluster)>,
    keepalive=120,
    ssl=True,
    ssl_params={'server_hostname':'da233fb2......2bf6c.s2.eu.hivemq.cloud'}
    )
    client.connect()
    return client


client = connectMQTT()

async def publish(topic, value):    # ensure than you have connection with client

  while True:
       print(topic)
       print(value)
       client.publish(topic, value)
       print("publish Done")
       await asyncio.sleep(1) 

# do any async work in main app ......

```

**Reply by nastiliano** [reply]
Hello,

Thanks, but for me is the same.
It doesn't work

Best regards.

**Comment by peterhinch** [comment]
The code looks OK to me. When you say it doesn't work, what happens? Please provide the traceback.

The advice from @2dof to test manually is good. I use `mosquitto_pub` and `mosquitto_sub` to prove I can access a broker before trying with `mqtt_as`.

**Reply by nastiliano** [reply]
Hello,

The message error is always the same:

    >>> Checking WiFi integrity.
    Got reliable connection
    Connecting to broker.
    Traceback (most recent call last):
      File "<stdin>", line 41, in <module>
      File "asyncio/core.py", line 1, in run
      File "asyncio/core.py", line 1, in run_until_complete
      File "asyncio/core.py", line 1, in run_until_complete
      File "<stdin>", line 26, in main
      File "mqtt_as.py", line 645, in connect
      File "mqtt_as.py", line 280, in _connect
    OSError: [Errno 113] ECONNABORTED

Thanks, best regards

**Reply by peterhinch** [reply]
It seems that the broker is rejecting the connection. I would definitely try with another client (`mosquitto_pub` and `mosquitto_sub` on a PC) to determine the reason.

**Reply by nastiliano** [reply]
Hello,

With my PC (using Python paho-mqtt) is working fine (configured as MQTT 5 version)

Simply:


        # "client MQTT" object
        self.client = paho.Client(client_id = "", userdata = None , protocol = paho.MQTTv5)

        # Credentials
        self.client.username_pw_set(".....", ".....")

        # Enabling TLS
        self.client.tls_set(tls_version = mqtt.client.ssl.PROTOCOL_TLS)

        # Trying a connection
        try:
            self.client.connect(".......s2.eu.hivemq.cloud" , 8883)
            self.client.disconnect()
            self.connection_state = True
        except:
            self.connection_state = False


Maybe is due to MQTT version? But HiveMQ accepts all versions of MQTT.... I don't know

I'm very lost :(


Thanks.

**Reply by zcattacz** [reply]
Make sure you are not using the same `user` or `client_id` on both instance. Kill all other instance connecting with the same credential, before testing.

**Reply by nastiliano** [reply]
Hi,

Absolutely impossible to connect to anything.... Even with public broker:


    from mqtt_as import MQTTClient, config
    import asyncio

    # Local configuration
    config['ssid'] = 'mySSID'  # Optional on ESP8266
    config['wifi_pw'] = 'mypass'
    config['server'] = 'mqtt.eclipse.org'  # Change to suit e.g. 'iot.eclipse.org'
    config['port'] = 1883

    def callback(topic, msg, retained):
        print((topic, msg, retained))

    async def conn_han(client):
        await client.subscribe('foo_topic', 1)

    async def main(client):
        await client.connect()
        n = 0
        while True:
            await asyncio.sleep(5)
            print('publish', n)
            # If WiFi is down the following will pause for the duration.
            await client.publish('result', '{}'.format(n), qos = 1)
            n += 1

    config['subs_cb'] = callback
    config['connect_coro'] = conn_han

    MQTTClient.DEBUG = True  # Optional: print diagnostic messages
    client = MQTTClient(config)
    try:
        asyncio.run(main(client))
    finally:
        client.close()  # Prevent LmacRxBlk:1 errors

Could you help me please??

    >>> Checking WiFi integrity.
    Got reliable connection
    Traceback (most recent call last):
      File "<stdin>", line 32, in <module>
      File "asyncio/core.py", line 1, in run
      File "asyncio/core.py", line 1, in run_until_complete
      File "asyncio/core.py", line 1, in run_until_complete
      File "<stdin>", line 17, in main
      File "mqtt_as.py", line 630, in connect
    OSError: -202

Thank you very much. Merry christmas.

**Reply by peterhinch** [reply]
It's failing on the DNS lookup:
```py
self._addr = socket.getaddrinfo(self.server, self.port)[0][-1]
```
`.server` is the entry in `config["server"]` and `.port` is 1883 unless you're using SSL in which case 8883. These are the standard ports for MQTT. Either the config entry is wrong or there is something amiss with DNS lookup on your network.

**Reply by nastiliano** [reply]
Ok, changing this it worked (public broker):

    # Local configuration
    config['ssid'] = 'mySSID'  # Optional on ESP8266
    config['wifi_pw'] = 'mypass'
    config['server'] = 'broker.hivemq.com'  # HiveMQ

But in private broker:

    # Local configuration
    config['ssid'] = 'mySSID'  # Optional on ESP8266
    config['wifi_pw'] = 'mypass'
    config['server'] = 'xxxxxxxxxxxxxxxxxxxxx.s2.eu.hivemq.cloud'  # Change to suit e.g. 'iot.eclipse.org'
    config['ssl'] = True
    config['user'] = 'user'
    config['password'] = 'pass'

Or even this:

    # Local configuration
    config['ssid'] = 'mySSID'  # Optional on ESP8266
    config['wifi_pw'] = 'mypass'
    config['server'] = 'xxxxxxxxxxxxxxxxxxxxx.s2.eu.hivemq.cloud'  # Change to suit e.g. 'iot.eclipse.org'
    config['ssl'] = True
    config['ssl_params'] = {"server_hostname": 'xxxxxxxxxxxxxxxxxxxxx.s2.eu.hivemq.cloud'}
    config['user'] = 'user'
    config['password'] = 'pass'


Fails, the error is always the same:
    
    Traceback (most recent call last):
    File "<stdin>", line 37, in <module>
    File "asyncio/core.py", line 1, in run
    File "asyncio/core.py", line 1, in run_until_complete
    File "asyncio/core.py", line 1, in run_until_complete
    File "<stdin>", line 22, in main
    File "mqtt_as.py", line 645, in connect
    File "mqtt_as.py", line 280, in _connect
    OSError: [Errno 113] ECONNABORTED

I think another parameter is missing...
Any ideas?

Thanks

**Reply by nastiliano** [reply]
Definitively,

Only works with public brokers...

I tried with "EMQX" too and the error is exactly the same.

Could you help me, please?

Thanks

**Reply by zcattacz** [reply]
If the server is kicking your device off, you need to read your broker service's documentation more carefully, it's not a client code issue. Make sure your client and broker setup is correct. Logon limit of identical credential is only one common limit of all possible limits.

**Reply by nastiliano** [reply]
Hello,

Finally after read a lot and a lot...

I installed "Mosquitto" in my own "VPS" and it works great, only using credentials without TLS certificates, because it's a nightmare.

Thank you very much and happy new year.
